### PR TITLE
Add `mapNonNullIndexed` API to `List` and `Set`

### PR DESCRIPTION
**Pull request by HT154**

### CODE REVIEW

File: pkl-core/src/test/files/LanguageSnippetTests/input/api/list.pkl [hunk 1]
```diff
@@ -332,6 +332,10 @@ examples {
     List(1, 2, 3).mapIndexed((i, n) -> i.isOdd && n.isEven)
   }
 
+  ["mapNonNullIndexed()"] {
+    List(1, 2, 3, 4).mapNonNullIndexed((i, n) -> if (n.isOdd) null else n * i)
```
Comment (bioball) [inline pkl-core/src/test/files/LanguageSnippetTests/input/api/list.pkl:336]: Can you add a test case that exercises null members?

Same for `Set`.

File: stdlib/base.pkl [hunk 2]
```diff
@@ -2441,6 +2441,16 @@ abstract external class Collection<out Element> extends Any {
   @AlsoKnownAs { names { "filterMap" } }
   abstract function mapNonNull<Result>(transform: (Element) -> Result): Collection<Result(this != null)>
 
+  /// Transforms this collection by applying [transform] to each element and removing resulting [null] elements.
+  ///
+  /// [transform] takes two arguments: the index of the element, and the element itself.
+  ///
+  /// Facts:
+  /// ```
+  /// List(1, 2, 3, 4).mapNonNullIndexed((i, n) -> if (n.isOdd) null else n * i) == List(2, 12)
```
Comment (bioball) [inline stdlib/base.pkl:2450]: Let's add another example here that includes a null member

**Review by bioball — CHANGES_REQUESTED** [review]
Thanks! Generally looks good, just two comments.

**Review by bioball — APPROVED** [review]
need to add the `@Since` annotation, but LGTM otherwise. Thanks!

**Comment by bioball** [comment]
LGTM on my wide, looking for a thumbs from @stackoverflow

**Review by stackoverflow — APPROVED** [review]
Looks good!
